read_write.go

```diff
@@ -156,53 +153,40 @@ func Encode(data interface{}, xmlAttributes []xml.Attr, ns string) ([]byte, erro
 	return xmlBytes, nil
 }
 
-// addNameSpace will check if a name space was passed and find the first tag
-// where it should be applied.
-// TODO: This should work on any depth and tag, including extensions. I guess
-// we must support multiple name spaces and aliases as well.
-func addNameSpace(document *xmltree.Element, alias string) {
-	if alias == "" {
-		return
+// addNameSpaceAlias will check each node/element in the XML tree and if the
+// node has an xml.Name.Space value set an alias will be created and then added
+// to all child nodes. The alias will only be setup for the root element.
+func addNameSpaceAlias(document *xmltree.Element, nsAdded bool) *xmltree.Element {
+	namespaceAliases := map[string]string{
+		types.NameSpaceDomain:   "domain",
+		types.NameSpaceHost:     "host",
+		types.NameSpaceContact:  "contact",
+		types.NameSpaceDNSSEC10: "sed",
+		types.NameSpaceDNSSEC11: "sec",
+		types.NameSpaceIIS12:    "iis",
 	}
 
-	childElement := document
-
-	// For default commands we shuld find the first element where we want to
-	// apply the name space alias. If we run out of children before the n:th tag
-	// is found we do nothing.
-	for range make([]int, nsCommandTagStartDepth) {
-		if len(childElement.Children) != 1 {
-			return
+	if document.Name.Space != "" {
+		alias, ok := namespaceAliases[document.Name.Space]
+		if !ok {
+			return nil
 		}
 
-		childElement = &childElement.Children[0]
-	}
-
-	xmlns := fmt.Sprintf("xmlns:%s", alias)
-	xmlnsValue := types.AliasToNameSpace(alias)
-
-	// Add the attribute xmlns with the appropreate value based on name space.
-	// TODO: Now we just support domain, contact and host.
-	childElement.SetAttr("", xmlns, xmlnsValue)
-
-	// Rename the local tag with the alias prefix, i.e. 'create' >
-	// 'domain:create'.
-	// TODO: Why can we not set element.Name.Space?
-	childElement.Name.Local = fmt.Sprintf("%s:%s", alias, childElement.Name.Local)
+		if !nsAdded {
+			xmlns := fmt.Sprintf("xmlns:%s", alias)
```

Comment:
Don't know if this is me or one of all the issues linked but this is super hackish. Also there's no way to remove the default added attribute `xmlns` so the tag will have both `obj:xmlns` and `xmlns` with the same value.